[flang][OpenMP] Convert DataSharingProcessor to omp::Clause

kparzysz opened the pull request:
[Clause representation 6/6]

skatrak commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -1384,30 +1386,28 @@ void DataSharingProcessor::copyLastPrivateSymbol(
-  for (const Fortran::parser::OmpObject &ompObject : ompObjectList.v) {
-    Fortran::semantics::Symbol *sym = getOmpObjectSymbol(ompObject);
+  for (const omp::Object &object : objects) {
+    Fortran::semantics::Symbol *sym = object.sym;
```
Nit: I think `object.sym` can be passed directly to the call below now that this becomes a trivial statement.

kparzysz replied:
Done.

skatrak reviewed (commented):
LGTM, just small comments.

kparzysz committed:
[flang][OpenMP] Convert DataSharingProcessor to omp::Clause

[Clause representation 6/6]

This can be committed out of order, because the prerequitites have already
been merged.

kiranchandramohan commented on flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:
```diff
@@ -135,138 +133,135 @@ void DataSharingProcessor::insertBarrier() {
 void DataSharingProcessor::insertLastPrivateCompare(mlir::Operation *op) {
   bool cmpCreated = false;
   mlir::OpBuilder::InsertPoint localInsPt = firOpBuilder.saveInsertionPoint();
-  for (const Fortran::parser::OmpClause &clause : opClauseList.v) {
```
Are there code changes here or is it only  the replacement of `parser::OmpClause` with `omp::Clause`?

kparzysz replied:
The code functionality is unchanged.  One reason for the seemingly large differences in the diff is that I replaced the `if (get_if<LastPrivate>)` with `if (!get_if<Lastprivate>) continue;` to reduce indentation.

kiranchandramohan reviewed (approved):
LG.

kparzysz commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -1278,14 +1278,15 @@ class DataSharingProcessor {
   llvm::SetVector<const Fortran::semantics::Symbol *> symbolsInNestedRegions;
   llvm::SetVector<const Fortran::semantics::Symbol *> symbolsInParentRegions;
   Fortran::lower::AbstractConverter &converter;
+  Fortran::semantics::SemanticsContext &semaCtx;
```
Removed.
